### ISSUE TITLE
xmonad --restart: some config changes ignored due to passed state

### ISSUE THREAD

**Issue by SOwOphie** [xmonad/xmonad#125]
### Problem Description

My setup includes a theme switcher script to switch between day and night mode. It basically replaces a few RGB values and then restarts everything. This method is incompatible with `xmonad-contrib` layouts that contain colour values in their configuration (such as `X.L.ShowWName` and `X.L.SimpleDecoration`), as these retain their internal state across a `--restart`, making config changes have no effect.

### Reproduction Steps

* `xmonad.hs`:
```haskell
module Main (main) where

import XMonad
import XMonad.Layout.SimpleDecoration

main :: IO ()
main = xmonad def { layoutHook = simpleDeco
                                    shrinkText
                                    (def {activeTextColor = "#FF0000"})
                                    (layoutHook def)
                  }

```

* `mod-q` ==> red label on active window
* change `activeTextColor` to `#00FF00`
* `mod-q` ==> still a red label on active window
* `ln -sf /dev/null ~/.xmonad/xmonad.state`
* change `activeTextColor` to `#0000FF`
* `mod-q` ==> label now blue

### Checklist

  - [x] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - [x] I tested my configuration with [xmonad-testing](https://github.com/xmonad/xmonad-testing)

**Edit:** I wasn't entirely certain whether this is an `xmonad` or an `xmonad-contrib` issue, in the end `xmonad` won by a narrow margin.

**Comment by geekosaur** [comment]
mod-shift-space. Or are you asserting that saved state, which xmonad has done since its earliest versions, is always the wrong thing to do?

**Comment by SOwOphie** [comment]
No, that's not the case. As I've spent this day with `~/.xmonad/xmonad.state` pointing to `/dev/null`, I can assure you that I appreciate this feature =)

Also, a complete mod-shift-q restart is not what I'm looking for here. Having to re-login just to update my window decoration's colour scheme seems like a waste to me, especially considering the functionality for live updates is already 99% there.

Take [X.L.ShowWName.ShowWName](http://hackage.haskell.org/package/xmonad-contrib-0.13/docs/src/XMonad-Layout-ShowWName.html#ShowWName) for example, the type contains a state (which should be retained across restarts), a config (which should be re-read), and a magic `Bool`. Is there some way to tell the layout to re-initialize its config after the reload? If not, how hard would it be to implement?

From my position of relative ignorance, I see three approaches to this problem:
* Use the existing message bus to send a `ReloadPlease newLayout` containing the layout's _new_ default state. Existing layouts can simply ignore this message to default to the old behaviour, the modules in question could trivially update the relevant parts of their config from that.
* Provide some kind of static storage outside the layout tree which resets with every reload, and make the modules in question use that. Maybe a bit too weird?
* If there are no viable options to fix this, put some hint like "please restart xmonad to apply colour changes" in the modules' doc. While that doesn't solve my particular problem, it should at least help new users trying to get their colour scheme right (been there, done that).

**Comment by liskin** [comment]
mod-shift-space, not mod-shift-q.

**Comment by SOwOphie** [comment]
While that is a step in the right direction, it also doesn't solve the problem entirely:
* The snippet behind mod-shift-space only fixes the current workspace so I'd have to map it over all of them, but more importantly:
* It also resets the window positioning. For `X.L.BinarySpacePartition`, this results in a heap of progressively tinier windows in the top left corner. Re-arranging them manually is a huge pain.

**Comment by SOwOphie** [comment]
Alright, I've tinkered around a bit with this problem and I think I can see the solution. Please take a look at SirBoonami/xmonad@4afcd67517044cba563d92554748c873d82f4f10 and SirBoonami/xmonad-contrib@195ae27aa8c3c0e3d17a091b0368249fda8cf585.

* I've added `handleReload :: layout a -> layout a -> layout a` to `LayoutClass`. The first parameter is the old state read from the state file, the second is the new default from the user supplied config. The default implementation, `handleReload = const`, mimics the current behaviour.
* This function has been hooked up into `readStateFile` using some helpers. If the parsed state is of the same type as the new one, `handleReload` is called. Otherwise, the new layout is used. This also mimics the old behaviour.
* xmonad-contrib has received some example implementations of `handleReload`.
* Most notably `X.L.LayoutModifier.LayoutModifier` has received a new function `handleReloadMod :: m a -> m a -> m a`, analogous to `handleReload`.

Now my problem is coming up with a definition as to which properties exactly should be reloaded for a given layout.

* Internal state such as window placement should obviously be kept.
* Internal configuration such as colour scheme should obviously be reloaded.
* But what about dynamic properties such as the ratio of `X.L.Spiral`? If the user is trying to find a good setting via trial-and-error, replacing the ratio would be a good thing. But if he's using a theme script after manually shaping the spiral via keybinds, he'd probably be pissed at the reset. I'd go for "keep it if it can be modified via messages, there's always mod-shift-space", what do you think?

**Comment by alessandro-molari** [comment]
Also keybidings changes aren't applied with `xmonad --restart`.

**Comment by liskin** [comment]
@SirBoonami You might also want to have a look at my attempt from 2009 to solve a similar issue: https://mail.haskell.org/pipermail/xmonad/2009-December/009308.html. The code is these days available at https://github.com/liskin/liskin-xmonad-conf/blob/a0c703c2e2af69bc228145112eb9ee7b9e40a26f/lib/XMonad/Layout/FlexibleRead.hs, https://github.com/liskin/liskin-xmonad-conf/tree/a0c703c2e2af69bc228145112eb9ee7b9e40a26f/lib/XMonad/Util. I'm not sure if there's example usage but the git history goes all the way to 2009 so it might be.

Regarding your question of which properties exactly should be reloaded for a given layout, I think you may avoid this question entirely and do a 3-way diff/merge instead:

1. before restarting, serialize not just the current layout of every workspace, but also the default/initial layout
1. after restart, combine the three together and figure out where the old initial layout differs from the new initial layout and apply these changes to the layouts of workspaces

**Comment by luc-tielen** [comment]
I had a similar problem when going from 9 workspaces to 4.. the last 5 kept showing up due to xmonad.state keeping track of all workspaces (including older/deleted ones).
Fixed it by making xmonad.state point to /dev/null with symbolic link, then restart and unlink.

**Comment by mwatts15** [comment]
Is this still relevant? It doesn't seem like XMonad pays attention to any `xmonad.state` file.

**Comment by liskin** [comment]
@mwatts15 Yes, it's still relevant. Changing parameters of layouts like themes or various constants still isn't applied automatically without mod-shift-space.
